ksqldb-engine/src/main/java/io/confluent/ksql/engine/TransientQueryCleanupService.java

```diff
@@ -0,0 +1,312 @@
+import java.util.concurrent.Callable;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.apache.kafka.streams.StreamsConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TransientQueryCleanupService extends AbstractScheduledService {
+  private static final Logger LOG = LoggerFactory.getLogger(TransientQueryCleanupService.class);
+  private final Pattern transientPattern;
+  private static final Pattern LEAKED_TOPIC_PREFIX_PATTERN =
+          Pattern.compile("(?i).*transient_.*_[0-9]\\d*_[0-9]\\d*");
+
+  private final BlockingQueue<Callable<Boolean>> stateDirsCleanupTasks;
+  private final BlockingQueue<Callable<Boolean>> topicsCleanupTasks;
+  private Set<String> localCommandsQueryAppIds;
+  private boolean isLocalCommandsInitialized;
+  private boolean isLocalCommandsProcessed;
+  private QueryRegistry queryRegistry;
+  private final String stateDir;
+  private final ServiceContext serviceContext;
+  private final int initialDelay;
+  private final int intervalPeriod;
+
+
+  public TransientQueryCleanupService(final ServiceContext serviceContext,
+                                      final KsqlConfig ksqlConfig) {
+    this.transientPattern = Pattern.compile(
+            ksqlConfig.getString(
+                    KsqlConfig.KSQL_TRANSIENT_QUERY_NAME_PREFIX_CONFIG));
+
+    this.initialDelay = ksqlConfig.getInt(
+            KsqlConfig.KSQL_TRANSIENT_QUERY_CLEANUP_SERVICE_INITIAL_DELAY_SECONDS);
+
+    this.intervalPeriod = ksqlConfig.getInt(
+            KsqlConfig.KSQL_TRANSIENT_QUERY_CLEANUP_SERVICE_PERIOD_SECONDS);
+
+    topicsCleanupTasks = new LinkedBlockingDeque<>();
+    stateDirsCleanupTasks = new LinkedBlockingDeque<>();
+
+    this.stateDir = ksqlConfig.getKsqlStreamConfigProps()
+            .getOrDefault(
+                    StreamsConfig.STATE_DIR_CONFIG,
+                    StreamsConfig.configDef()
+                            .defaultValues()
+                            .get(StreamsConfig.STATE_DIR_CONFIG))
+            .toString();
+
+    this.serviceContext = serviceContext;
+    this.isLocalCommandsInitialized = false;
+    this.isLocalCommandsProcessed = false;
+  }
+
+  @Override
+  protected void runOneIteration() {
+    if (isLocalCommandsInitialized && !isLocalCommandsProcessed) {
+      LOG.info("Adding LocalCommands to TransientQueryCleanupService.");
+      localCommandsQueryAppIds.forEach(id -> {
+        addTopicCleanupTask(new TransientQueryTopicCleanupTask(serviceContext, id));
+        addStateCleanupTask(new TransientQueryStateCleanupTask(id, stateDir));
+      });
+      isLocalCommandsProcessed = true;
+    }
+
+    LOG.info("Starting cleanup for leaked resources.");
+
+    topicsCleanupTasks.addAll(findPossiblyLeakedTransientTopics());
+    stateDirsCleanupTasks.addAll(findPossiblyLeakedStateDirs());
```

Comment:
So we enqueue all possible tasks here, and later, if we fail on a task we also re-enqueue it. So, on the next run, the same task will be in the queue twice. I don't think that's a problem, but on the other hand, since we have these two lines here, it doesn't seem like we strictly need those queues to persist between runs. We could also just freshly discover all potentially leaked resources from scratch every time, and if anything fails, oh well, we'll discover it again next time.

Reply:
>We could also just freshly discover all potentially leaked resources from scratch every time, and if anything fails, oh well, we'll discover it again next time.

I agree, the retry logic is unnecessarily complex. changed the logic to `just freshly discover all potentially leaked resources from scratch every time, and if anything fails, oh well, we'll discover it again next time.`